Use internal buffers for generated macro and template code

spaceotter opened the pull request:
Instead of generating code to a file, and having the user run c2ffi a second time with the additional code, this patch writes the extra code to a buffer and crams it through the parser after the main input file, eliminating the extra step and associated file I/O. In order to pull this off I ended up having to copy the clang ParseAST code just to change one line.

The PR also ignores macros with curly braces. I had noticed that such macros can sometimes cause valid constants to be dropped, and since the order of the generated constants isn't deterministic for some reason (??) you also have no idea which constants it will erase.

Please note that this breaks backwards compatibility of the command line interface, since the -T and -M options no longer need a path to output to.

spaceotter committed:
Reject macros with brace tokens

spaceotter committed:
Use a buffer for macro constants and parse immediately

To avoid the inconvenience of having to run the program twice, send the
generated constants for macros to a memory buffer and use a hack to
force the clang parser to accept the buffer as additional code. This
accomplishes the same thing with fewer steps from the user's
perspective, and without any file I/O.

spaceotter committed:
Only report errors that occurred in the input file

This distinguishes errors that occurred in the input file, which might
indicate a real problem, and errors that occurred in the generated
constants for macros. These are basically guaranteed to occur since it
just guesses the type and some macros contain something other than
constants.

Also cleared up the indentation from the last patch.

spaceotter committed:
Use found LLVM as hint to find Clang

This helps cmake find the clang installed in the same place as LLVM, so
hopefully the versions will match.

spaceotter committed:
Use a buffer for template specializations too

In order to make this work properly, I had to copy the entire ParseAST
file from clang just to change one line. The trick that was used with
the macros crashed on the template code, probably because internally
calling EnterMainSourceFile a second time resets some internal data.

spaceotter committed:
Update README and help output

The README and help output have been updated to reflect the changes to
macro & template handling.

spaceotter committed:
Change preamble of iparseast.* files

spaceotter committed:
Reuse the clang::Sema

This is necessary since the Sema stores some data that the template
specialization needs, destructing it was causing memory faults later on
in the template code. It still crashes on an assertion though.

spaceotter committed:
Dodge an assert failure because of lingering context

spaceotter committed:
Eliminate unnecessary changes to the main file

It doesn't seem to be necessary to change the main file ID for this hack.

spaceotter committed:
Fix indentation

This fixes indentation to match the rest of the project except in the
ParseAST code that was copied from clang.

spaceotter commented:
Well it doesn't crash for me anymore. I'm putting this up for consideration.

rpav commented:
This seems like another great improvement; I'm not sure anyone really uses `-T`, but `-M` is definitely used by autowrap and possibly others.  One possibility might be to retain `-M` for the time being as a "deprecated" parameter (warning, etc) and use `-m`.  Autowrap should be fairly easy to update, of course, but others might not. 

Which brings me to a tangential (rhetorical?) question; does anyone (or why does no one) have any kind of software-dependency-interest notification service for notifying users of breaking changes as such.

This patch may take me a few days to get to, but at a glance it looks like something that ought to be in.

spaceotter committed:
Move macro injection to a new option

The feature is made available as -m instead of replacing -M, which would
break reverse compatibility with downstream projects that use -M. Both
options can be used at the same time.

spaceotter committed:
Update help text

rpav commented:
This should have been merged but really c2ffi needs a serious rewrite, which as soon as I have a stable working dev environment again is likely to happen quickly.
